File: benchmarks/src/lua/closure_chain.lua
local function make_counter(start)
    local c = start
    return function()
        c = c + 1
        return c
    end
end

local N = 150000
local acc = 0
local start = os.clock()
for i = 1, N do
    local ctr = make_counter((i - 1) % 1000)
    acc = (acc + ctr() + ctr() + ctr()) % 1000003
end
local finish = os.clock()
print("Result: " .. acc)
print(string.format("Time: %.0fms", (finish - start) * 1000))

File: benchmarks/src/perl/closure_chain.pl
use Time::HiRes qw(time);

sub make_counter {
    my $start = shift;
    my $c = $start;
    return sub { $c = $c + 1; return $c; };
}

my $N = 150000;
my $acc = 0;
my $start = time();
for (my $i = 0; $i < $N; $i++) {
    my $ctr = make_counter($i % 1000);
    $acc = ($acc + $ctr->() + $ctr->() + $ctr->()) % 1000003;
}
my $end = time();
my $ms = ($end - $start) * 1000;
print "Result: $acc\n";
printf "Time: %.0fms\n", $ms;
